File: Storyteller/STSelectSceneElementViewController.m
//
//  STSelectSceneElementViewController.m
//  Storyteller
//
//  Created by Ryan Britt on 7/5/14.
//  Copyright (c) 2014 Ryan Britt. All rights reserved.
//

#import "STSelectSceneElementViewController.h"
#import "STEditSceneViewController.h"
#import <UIView+DragDrop.h>

@interface STSelectSceneElementViewController ()
@property (strong, nonatomic) STEditSceneViewController *editSceneVC;
@property (strong, nonatomic) NSArray *sceneElementImageArray;
@property SelectedSceneElementType currentSceneElementType;
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@end

@implementation STSelectSceneElementViewController

-(NSArray *)sceneElementImageArray
{
    if (!_sceneElementImageArray)
    {
        _sceneElementImageArray = [[NSArray alloc]init];
    }
    return  _sceneElementImageArray;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.sceneElementImageArray = @[@"Actor",@"Actor 2", @"Actor 3"];    // Do any additional setup after loading the view.
    self.currentSceneElementType = SelectedSceneElementTypeActor;
    self.collectionView.backgroundColor = [UIColor groupTableViewBackgroundColor];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)addTextButton:(id)sender
{
    [self.editSceneDelegate showElementSelect:sender];
    [self.editSceneDelegate addTextButton:sender];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
    
    UIImage *image = [UIImage imageNamed:self.sceneElementImageArray[indexPath.row]];
    switch (self.currentSceneElementType)
    {
        case SelectedSceneElementTypeActor:
        case SelectedSceneElementTypeEnvironment:
        case SelectedSceneElementTypeObject:
            cell.backgroundView = [[UIImageView alloc]initWithImage:image];
            break;
        case SelectedSceneElementTypeMedia:
        {
            UITextView *textView = [[UITextView alloc]initWithFrame:CGRectMake(0, 0, 100, 200)];
            [textView setText:@"Text Box"];
            textView.editable = NO;
            textView.userInteractionEnabled = NO;
            textView.backgroundColor = [UIColor grayColor];
            cell.backgroundView = textView;
        }
            break;
        default:
            break;
    }

    return cell;
    
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.sceneElementImageArray.count;
}

-(void)tabBar:(UITabBar *)tabBar didSelectItem:(UITabBarItem *)item
{
    if ([item.title isEqualToString:@"Actor"])
    {
        self.currentSceneElementType = SelectedSceneElementTypeActor;
        self.sceneElementImageArray = @[@"Actor",@"Actor 2", @"Actor 3"];
        [self.collectionView reloadData];
    }
    else if ([item.title isEqualToString:@"Environment"])
    {
        self.currentSceneElementType = SelectedSceneElementTypeEnvironment;
        self.sceneElementImageArray = @[@"Environment", @"Environment 2", @"Environment3"];
        [self.collectionView reloadData];
    }
    else if ([item.title isEqualToString:@"Object"])
    {
        self.currentSceneElementType = SelectedSceneElementTypeObject;
        self.sceneElementImageArray = @[@"Object", @"Object 2", @"Object 3"];
        [self.collectionView reloadData];
    }
    else if ([item.title isEqualToString:@"Media"])
    {
        self.currentSceneElementType = SelectedSceneElementTypeMedia;
        self.sceneElementImageArray = @[@"textfield"];
        [self.collectionView reloadData];
    }
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *currentCell = [self.collectionView cellForItemAtIndexPath:indexPath];
    UIImage *currentImage;
    if ([currentCell.backgroundView isMemberOfClass:[UIImageView class]])
    {
        currentImage = ((UIImageView *)currentCell.backgroundView).image;
    }

    switch (self.currentSceneElementType)
    {
        case SelectedSceneElementTypeActor:
            [self.editSceneDelegate showElementSelect:nil];
            [self.editSceneDelegate addActorSceneElementWithImage:currentImage];
            break;
        case SelectedSceneElementTypeEnvironment:
            [self.editSceneDelegate showElementSelect:nil];
            [self.editSceneDelegate addEnvironmentSceneElementWithImage:currentImage];
            break;
        case SelectedSceneElementTypeObject:
            [self.editSceneDelegate showElementSelect:nil];
            [self.editSceneDelegate addObjectSceneElementWithImage:currentImage];
            break;
        case SelectedSceneElementTypeMedia:
            [self.editSceneDelegate showElementSelect:nil];
            [self.editSceneDelegate addTextButton:nil];
            break;
            
        default:
            break;
    }
    
}
@end

File: Storyteller/STSelectSceneElementViewController.swift
//
//  STSelectSceneElementViewController
//  Storyteller
//
//  Created by Ryan Britt on 7/17/14.
//  Copyright (c) 2014 Ryan Britt. All rights reserved.
//
// refactor to: STSelectSceneElementViewController

import UIKit

class STSelectSceneElementViewController:UIViewController
{

    enum SelectedSceneElementType
    {
        case Actor
        case Environment
        case Object
        case Text
    }
    
    
    //class properties
    
    //Any or all of these could be empty in which case there are no elements of a particular type.
    var sceneElementPathArray:NSArray = []
    var actorElementPathArray:NSArray = []
    var environmentElementPathArray:NSArray = []
    var objectElementPathArray:NSArray = []

    var currentSceneElementType = SelectedSceneElementType.Actor
    @objc var sceneManagementDelegate:SceneManagementDelegate?

    @IBOutlet weak var collectionView: UICollectionView!
    @IBOutlet weak var tabBar: UITabBar!
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        collectionView.backgroundColor = UIColor.groupTableViewBackground
        
        if let uitabBarItems = tabBar.items
        {
            tabBar.selectedItem = uitabBarItems[0] as UITabBarItem;
        }
        
        //find the path to the resource dictionary
        let resourcePath = Bundle.main.path(forResource: "ImageResourceNames", ofType: "plist")
        //initialize the resource dictionary
        let resourcePathDictionary = NSDictionary(contentsOfFile: resourcePath!)
        
        actorElementPathArray = (resourcePathDictionary?.object(forKey: "Actor") as! NSArray)
        environmentElementPathArray = (resourcePathDictionary?.object(forKey: "Environment") as! NSArray)
        objectElementPathArray = (resourcePathDictionary?.object(forKey: "Object") as! NSArray)
        
        
//        let allkeys = resourcePathDictionary.allKeys as [String]
//        //{// if let statement needed above for Xcode 6.1 beta. Not sure if permanent swift language change or not.
//            for key in allkeys
//            {
//                //put the appropriate objects into each array
//                let value = resourcePathDictionary.objectForKey(key) as [String]
//                switch key
//                {
//                case "Actor":
//                    actorElementPathArray = value
//                case "Environment":
//                    environmentElementPathArray = value
//                case "Object":
//                    objectElementPathArray = value
//                default:
//                    break
//                }
//
//            }
//        //}
        
        sceneElementPathArray = actorElementPathArray


    }
    
}

extension STSelectSceneElementViewController:UICollectionViewDataSource
{
    func numberOfSectionsInCollectionView(collectionView: UICollectionView!) -> Int
    {
        return 1
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int
    {
        return sceneElementPathArray.count;
    }
    
    func collectionView(_ thisCollectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell
    {
        let cell:UICollectionViewCell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as UICollectionViewCell
        if sceneElementPathArray.count > 0
        {
            let currentImage = UIImage(named:sceneElementPathArray[indexPath.row] as! String)
            switch currentSceneElementType
                {
            case .Actor,.Environment,.Object:
                cell.backgroundView = UIImageView(image: currentImage)
            case .Text:
                let textView = UITextView() //UITextView(frame: CGRect(x: 0, y: 0, width: 50, height: 50))
                textView.text = "Text Box"
                textView.isEditable = false
                textView.isUserInteractionEnabled = false
                cell.backgroundView = textView
            }
            
        }
        return cell
        
    }
}

extension STSelectSceneElementViewController:UICollectionViewDelegate
{
    func collectionView(thisCollectionView: UICollectionView!,
               layout collectionViewLayout: UICollectionViewLayout!,
          sizeForItemAtIndexPath indexPath: NSIndexPath!) -> CGSize
        
    {
        //Assign a default size
        var size = CGSize(width:100, height:100)
        
        //if the scene elementpath contains a valid UIImage name, create it, and set the collectionViewItem size to its size.
        if let currentImage = UIImage(named: sceneElementPathArray[indexPath.row] as! String)
        {
            switch currentSceneElementType
            {
            case .Actor,.Environment,.Object:
                //Change size to match the image
                size = currentImage.size
            default:
                break
            }
        }
        return size
    }
    
    
    func collectionView(_ collectionView:UICollectionView, didSelectItemAt indexPath:IndexPath)
    {
        let cell = self.collectionView(collectionView, cellForItemAt: indexPath)
        var currentImage:UIImage?
        
        if let backgroundView = cell.backgroundView
        {
            if(backgroundView is UIImageView)
            {
                currentImage = (cell.backgroundView as! UIImageView).image
            }
        }
        
        //unwrap and check sceneManagementDelegate
        if let sceneDelegate = sceneManagementDelegate
        {
            switch currentSceneElementType
                {
            case .Actor:
                sceneDelegate.addSceneElement(with: currentImage, ofType: "Character")
            case .Environment:
                sceneDelegate.addSceneElement(with: currentImage, ofType: "Environment")
            case .Object:
                sceneDelegate.addSceneElement(with: currentImage, ofType: "Object")
            case .Text:
                sceneDelegate.addText()
            }
        }
        else {} //The edit Scene Delegate was not set before initiating this class. Create Error handling for outside caller.
        

        
    }

}

extension STSelectSceneElementViewController: UITabBarDelegate
{
    func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem)
    {
        if let title = item.title
        {
            switch title
            {
                case "Actor":
                    currentSceneElementType = .Actor
                    sceneElementPathArray = actorElementPathArray
                case "Environment":
                    currentSceneElementType = .Environment
                    sceneElementPathArray = environmentElementPathArray
                case "Object":
                    currentSceneElementType = .Object
                    sceneElementPathArray = objectElementPathArray
                case "Media":
                    currentSceneElementType = .Text
                    sceneElementPathArray = ["Text"]
            default:
                break
            }
        }
        
        collectionView.reloadData()
    }
}
